File: boj_codes/C++17/2098.cpp
// https://www.acmicpc.net/problem/2098
// First Written : 20210106
// Last Modified : 20210106

#include <cstdio>

#define MIN(X,Y) (((X) < (Y)) ? (X) : (Y))
#define MAX_VAL 1999999999

int power_2(int exp){
    int base = 1;
    for(int i = 0; i < exp; i++)
        base *= 2;
    return base;
}

int TSP(int cur_city, int visited, int** weight, int** dp, const int N){
    visited += power_2(cur_city);

    if(visited == power_2(N) - 1){
        if(weight[cur_city][0] == 0)
            return MAX_VAL;
        
        else
            return weight[cur_city][0];
        
    }

    if(dp[cur_city][visited] != 0)
        return dp[cur_city][visited];


    int min_weight = MAX_VAL;
    int cur_weight;
    int copied_visited = visited;

    for(int city = 0; city < N; city++){
        if((copied_visited % 2 == 0) && (weight[cur_city][city] != 0)){
            cur_weight = TSP(city, visited, weight, dp, N) + weight[cur_city][city];
            min_weight = MIN(cur_weight, min_weight);
        }
            
        copied_visited /= 2;
    }

    dp[cur_city][visited] = min_weight;
    return min_weight;
    
}

int main(){
    int N;
    scanf("%d", &N);

    int** weight = new int*[N];
    int** dp = new int*[N];

    for(int i = 0; i < N; i++){
        weight[i] = new int[N];
        dp[i] = new int[power_2(N)]();
    }

    for(int i = 0; i < N; i++)
        for(int j = 0; j < N; j++)
            scanf("%d", &weight[i][j]);

    printf("%d", TSP(0, 0, weight, dp, N));
    
    for(int i = 0; i < N; i++)
        delete [] dp[i], weight[i];

    delete [] dp, weight;

    return 0;
}

File: old_contents/boj_codes/Python3/2098.py
# https://www.acmicpc.net/problem/2098
# 20201103

import sys

MAX_VAL = 999999999

N_cities = int(sys.stdin.readline().rstrip())
N_visitied = (2 ** N_cities)

weights = [list(map(int, sys.stdin.readline().split())) for _ in range(N_cities)]
optimal = [[0] * N_visitied for _ in range(N_cities)]


def not_visited(visited):
    n_visited = []

    for i in range(N_cities):
        if visited % 2 == 0:
            n_visited.append(i)
        visited //= 2

    return n_visited


def TSP(current, visited):
    visited += (2 ** current)

    if visited == N_visitied - 1:
        if weights[current][0] == 0:
            return MAX_VAL
        else:
            return weights[current][0]

    else:
        if optimal[current][visited] != 0:
            return optimal[current][visited]

        min_weight = MAX_VAL

        for city in not_visited(visited):
            if weights[current][city] == 0:
                continue
            cur_weight = TSP(city, visited) + weights[current][city]

            if cur_weight < min_weight:
                min_weight = cur_weight

        optimal[current][visited] = min_weight
        return min_weight


print(TSP(0, 0))



